### PR TITLE
Sortable table

### PR DESCRIPTION
**Pull request by it-harrison**
## Chromatic

https://sortable-table--65a6e2ed2314f7b8f98609d8.chromatic.com

---

## Description
This PR makes the `va-table` sortable by means of a `sortable` prop.
Key points:
* the table can be sorted asc/desc by any column so long as the data in the column is numeric, date, or string
* the data projected into each table slot must be a string, not a component

Closes [2771](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2771)

## QA Checklist
- [ ] Component maintains 1:1 parity with design mocks
- [ ] Text is consistent with what's been provided in the mocks
- [ ] Component behaves as expected across breakpoints
- [ ] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [ ] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [ ] Tab order and focus state work as expected
- [ ] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [ ] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [ ] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Screenshots

Screenshot 2024-10-07 at 7 29 44 PM

## Acceptance criteria
- [ ] QA checklist has been completed
- [ ] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

**Comment by rsmithadhoc** [comment]
> * The screen reader announces the sort buttons on focus as "table groups" and the instructions on how to interact are not typical for a button and don't work. I think the buttons should have more descriptive instructions and get announced as buttons. @rsmithadhoc might be able to help better on that.

I believe those will need to be wrapped in a `<button>`. This should also have a `title` attribute with text saying "Click to sort by (column name) in (ascending/descending) order", that will read it out on a screen reader and also show a tooltip on hover.

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (jamigibbs) [inline packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx:52]: If `sortdir` and `sortindex` are public properties, can we add documentation comments for them to explain how they work? Would it also be valuable to add a story that demonstrates them?

**Comment by jamigibbs** [comment]
> Headings disappear on smaller screens, this is what the behavior in prod is currently. It seems like you should be able to sort table data on all screens, especially smaller screens where there is less real-estate and you want to be able to find things faster, but that is not currently possible. This may be a conversation to have with USWDS and design

@powellkerry Is what you're describing the same thing as this issue? 

- https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/3327

**Comment by it-harrison** [comment]
> https://sortable-table--65a6e2ed2314f7b8f98609d8.chromatic.com

@powellkerry @jamigibbs good call on the disappearing headers; I pushed a commit so that they don't disappear for a sortable table

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx]: @jamigibbs - good call - they don't need to be props, I have handled this using data- attrs and State in a refactor

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx:73]: @jamigibbs it should only be useful internally

File: packages/web-components/src/components/va-table/sort/alpha.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/sort/alpha.tsx:6]: @jamigibbs fixed!

File: packages/web-components/src/components/va-table/sort/constants.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/sort/constants.tsx]: @jamigibbs I wondered this too - I moved the constants to higher level utils folder because there is a chance of some other component using them

File: packages/web-components/src/components/va-table/sort/date.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/sort/date.tsx:22]: @jamigibbs - fixed!

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.scss
Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.scss]: @jamigibbs - yes there are! fixed!

**Comment by it-harrison** [comment]
> > * The screen reader announces the sort buttons on focus as "table groups" and the instructions on how to interact are not typical for a button and don't work. I think the buttons should have more descriptive instructions and get announced as buttons. @rsmithadhoc might be able to help better on that.
> 
> I believe those will need to be wrapped in a `<button>`. This should also have a `title` attribute with text saying "Click to sort by (column name) in (ascending/descending) order", that will read it out on a screen reader and also show a tooltip on hover.

@rsmithadhoc instead of a button I used a span with a `role="button"` so that I didn't have to style the button - VoiceOver was the same but want to make sure this is okay.

**Comment by rsmithadhoc** [comment]
> > > * The screen reader announces the sort buttons on focus as "table groups" and the instructions on how to interact are not typical for a button and don't work. I think the buttons should have more descriptive instructions and get announced as buttons. @rsmithadhoc might be able to help better on that.
> > 
> > 
> > I believe those will need to be wrapped in a `<button>`. This should also have a `title` attribute with text saying "Click to sort by (column name) in (ascending/descending) order", that will read it out on a screen reader and also show a tooltip on hover.
> 
> @rsmithadhoc instead of a button I used a span with a `role="button"` so that I didn't have to style the button - VoiceOver was the same but want to make sure this is okay.

It should be okay functionality-wise, Andrew ran into issues recently with linting and explicitly defining a `role`, so that linting rule had to be disabled. I believe the outcome was that the rule had to be disabled in a handful of `vets-website` tests, but ideally to remove that in the future and use the actual element. I'm not sure if we would want to expand on those one-off rule disables if we don't have to. But just know you might see issues in `vets-website`. Here's the Slack thread on that: https://dsva.slack.com/archives/G01BJ3ESXL4/p1727383688504589?thread_ts=1727380266.191099&cid=G01BJ3ESXL4

I'll give my approval, I also have https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2772 to give it a full a11y review as well.

**Review by rsmithadhoc — APPROVED** [review]

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (jamigibbs) [inline packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx:111]: I think [USWDS](https://designsystem.digital.gov/components/table/) is using a `button` with styling classes. Can we do that too here instead of using a `span`?

Screenshot 2024-10-15 at 11 27 36 AM

**Review by jamigibbs — COMMENTED** [review]
Have we tested how this behaves with [pagination](https://sortable-table--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/uswds-va-table--docs#with-pagination)?

File: packages/web-components/src/components/va-table/va-table.tsx
Comment (jamigibbs) [inline packages/web-components/src/components/va-table/va-table.tsx:46]: Have we tested how this behaves with [pagination](https://sortable-table--65a6e2ed2314f7b8f98609d8.chromatic.com/?path=/docs/uswds-va-table--docs#with-pagination)?

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx]: @jamigibbs - I switched to button but had to style it since the styles from uswds diverge from ours here

File: packages/web-components/src/components/va-table/va-table.tsx
Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table.tsx:46]: @jamigibbs there was a bug (not actually related to pagination but revealed in the pagination example) that is now fixed

**Comment by it-harrison** [comment]
@rsmithadhoc heads up - I went with buttons in the table header ths.

File: packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx
Comment (jamigibbs) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx]: @department-of-veterans-affairs/platform-design-system-designers Are we aware that there is a design divergence between USWDS table and our Figma?

Comment (it-harrison) [reply packages/web-components/src/components/va-table/va-table-inner/va-table-inner.tsx]: @jamigibbs - I have keyed Dan into this and he will be working on a solution going forward

**Review by jamigibbs — APPROVED** [review]
Excellent work, @it-harrison! 

If there is any time to add tests for the [sort utility methods](https://github.com/department-of-veterans-affairs/component-library/tree/07f83b726babde36450da277805456f9a4507608/packages/web-components/src/components/va-table/sort), there might be an opportunity to do that similar to how [va-telephone has unit tests](https://github.com/department-of-veterans-affairs/component-library/blob/main/packages/web-components/src/components/va-telephone/test/va-telephone.spec.tsx).

And with there being visual changes, let's try and have a design approval before merging too. 🙏🏼

**Comment by it-harrison** [comment]
> https://sortable-table--65a6e2ed2314f7b8f98609d8.chromatic.com

@jamigibbs sorting utils tests added. reaching out to design for review

**Comment by it-harrison** [comment]
@danbrady @LWWright7 @babsdenney as per a suggestion from @jamigibbs this PR will wait to merge until one of you has had time to review it for visual changes - thanks!

**Comment by danbrady** [comment]
Sorry for the delay in review, @it-harrison. I'm a bit swamped. Quick question on something I notice... How come the sorted icons are different between USWDS and VADS?

USWDS:
Screenshot 2024-10-22 at 2 59 43 PM

VADS:
Screenshot 2024-10-22 at 2 59 23 PM

**Comment by it-harrison** [comment]
> Sorry for the delay in review, @it-harrison. I'm a bit swamped. Quick question on something I notice... How come the sorted icons are different between USWDS and VADS?
> 
> USWDS: Screenshot 2024-10-22 at 2 59 43 PM
> 
> VADS: Screenshot 2024-10-22 at 2 59 23 PM

@danbrady no worries and good call - I used `va-icon` whereas uswds has their own icon set. I assumed that we should use `va-icon` (I think we have done so in other components that diverge in this way from uwswds) but I could be wrong.

**Comment by danbrady** [comment]
I see why you did that, Ian. That makes sense to me. 

Maybe the USWDS uses that slim arrow instead because it looks closer to the default sort icon. Whereas the Material/VADS arrows are downright *chonky*. However, the USWDS should probably include those slim arrows in their icon set.


Here's three possible ideas for a way forward:
1. Keep the icons as we have it now (not aligned with USWDS) using these: 
image

2. Use the images/unofficial icons that USWDS currently uses and not va-icon (and submit an issue to them):
image
 
3. Add a new, similar style sortable icon from Material to match the style of the arrows in va-icon:
image

@humancompanion-usds, your thoughts?

@it-harrison The rest of the design looks good. I don't think it's necessary to stop merging this based on the icon alone. We can create a follow up ticket, if needed. Thanks!

**Review by danbrady — APPROVED** [review]

**Pull request merged**
